src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,372 @@
+                  QVariant(), true ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink(
+                  QStringLiteral( "OUTPUT" ),
+                  QObject::tr( "Extracted labels" ),
+                  QgsProcessing::TypeVectorPoint ) );
+}
+
+QString QgsExtractLabelsAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Extract label information (position, font details, etc.) from a rendered map at a given extent and scale." );
+}
+
+QString QgsExtractLabelsAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
+                      "Extracted label information include its position (served as point geometries), the associated layer name and feature ID, as well as label text, rotation, alignment, and font details." );
+}
+
+QgsExtractLabelsAlgorithm *QgsExtractLabelsAlgorithm::createInstance() const
+{
+  return new QgsExtractLabelsAlgorithm();
+}
+
+class ExtractLabelSink : public QgsLabelSink
+{
+  public:
+    ExtractLabelSink( QMap<QString, QString> *mapLayerNames, QgsProcessingFeedback *feedback )
+      : mMapLayerNames( mapLayerNames )
+      , mFeedback( feedback )
+    {
+    };
+
+    void drawLabel( const QString &layerId, QgsRenderContext &context, pal::LabelPosition *label, const QgsPalLayerSettings &settings ) override
+    {
+      const QgsFeatureId fid = label->getFeaturePart()->featureId();
+      if ( settings.placement == QgsPalLayerSettings::Curved ||
+           settings.placement == QgsPalLayerSettings::PerimeterCurved )
+      {
+        mFeedback->pushWarning( QObject::tr( "Curved placement not supported, skipping label for FID %1 from layer %2" ).arg( fid ).arg( layerId ) );
+        return;
+      }
+
+      QgsTextLabelFeature *labelFeature = dynamic_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
+      QgsPalLayerSettings labelSettings( settings );
+      const QMap< QgsPalLayerSettings::Property, QVariant > &dataDefinedValues = labelFeature->dataDefinedValues();
+
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::MultiLineWrapChar ) )
+      {
+        labelSettings.wrapChar = dataDefinedValues.value( QgsPalLayerSettings::MultiLineWrapChar ).toString();
+      }
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::AutoWrapLength ) )
+      {
+        labelSettings.autoWrapLength = dataDefinedValues.value( QgsPalLayerSettings::AutoWrapLength ).toInt();
+      }
+      const QString labelText = QgsPalLabeling::splitToLines( labelFeature->text( -1 ),
+                                labelSettings.wrapChar,
+                                labelSettings.autoWrapLength,
+                                labelSettings.useMaxLineLengthForAutoWrap ).join( "\n" );
```

Comment:
Could this logic be moved to QgsTextLabelFeature?

Reply:
@nyalldawson , I'm unsure, see https://github.com/qgis/QGIS/blob/master/src/core/labeling/qgsvectorlayerlabelprovider.cpp#L659 . If we change that, we'd have to apply changes in other parts of the code too. 

One thing this work revealed is how much improvements we could do in the overall labeling code workflow, but maybe that's better done in a separate PR.